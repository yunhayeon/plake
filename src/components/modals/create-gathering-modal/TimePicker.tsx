import dayjs from "dayjs";
import { useState } from "react";

import { Button } from "@/components/ui/Button";

const baseStyle =
  "m-[10px] flex h-[100px] md:h-[300px] flex-col gap-[10px] overflow-y-scroll pl-3";

interface ITimePickerProps {
  selectedDate?: Date;
  setValue?: (value: Date) => void;
}

const TimePicker = ({ selectedDate, setValue }: ITimePickerProps) => {
  const [selectedHour, setSelectedHour] = useState<number>(12);
  const [selectedMinute, setSelectedMinute] = useState<number>(0);
  const [selectedAmPm, setSelectedAmPm] = useState<string>("PM");

  const updateDateTime = (
    hour: number = selectedHour,
    minute: number = selectedMinute,
    ampm: string = selectedAmPm,
  ) => {
    if (!selectedDate || !setValue) return;

    let adjustedHour = hour;

    if (ampm === "PM" && hour !== 12) {
      adjustedHour = hour + 12;
    } else if (ampm === "AM" && hour === 12) {
      adjustedHour = 0;
    }

    const formattedDateTime = new Date(
      dayjs(selectedDate)
        .hour(adjustedHour)
        .minute(minute)
        .format("YYYY-MM-DD hh:mm A"),
    );

    setValue(formattedDateTime);
  };

  return (
    <section className="flex gap-2 divide-x divide-gray-200">
      <div className={baseStyle}>
        {Array.from({ length: 12 }).map((_, index) => {
          const hour = index === 0 ? 12 : index;
          return (
            <Button
              variant={selectedHour === hour ? "purple" : "ghost"}
              key={index}
              onClick={() => {
                setSelectedHour(hour);
                updateDateTime(hour);
              }}
            >
              {hour.toString().padStart(2, "0")}
            </Button>
          );
        })}
      </div>
      <div className={baseStyle}>
        {Array.from({ length: 12 }, (_, index) => index * 5).map(minute => (
          <Button
            variant={selectedMinute === minute ? "purple" : "ghost"}
            key={minute}
            onClick={() => {
              setSelectedMinute(minute);
              updateDateTime(selectedHour, minute);
            }}
          >
            {minute.toString().padStart(2, "0")}
          </Button>
        ))}
      </div>
      <div className={baseStyle}>
        <Button
          variant={selectedAmPm === "PM" ? "purple" : "ghost"}
          onClick={() => {
            setSelectedAmPm("PM");
            updateDateTime(selectedHour, selectedMinute, "PM");
          }}
        >
          {"PM"}
        </Button>
        <Button
          variant={selectedAmPm === "AM" ? "purple" : "ghost"}
          onClick={() => {
            setSelectedAmPm("AM");
            updateDateTime(selectedHour, selectedMinute, "AM");
          }}
        >
          {"AM"}
        </Button>
      </div>
    </section>
  );
};

export default TimePicker;
