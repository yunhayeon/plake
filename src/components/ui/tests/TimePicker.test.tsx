import { fireEvent, render, screen } from "@testing-library/react";

import TimePicker from "../TimePicker";

describe("TimePicker 컴포넌트 테스트", () => {
  const mockSetValue = jest.fn();
  const mockSelectedDate = new Date("2024-03-20");

  describe("렌더링 테스트", () => {
    it("기본값이 제대로 렌더링되어야 함", () => {
      render(
        <TimePicker selectedDate={mockSelectedDate} setValue={mockSetValue} />,
      );

      expect(screen.getByLabelText("12시")).toMatchSnapshot();
      expect(screen.getByLabelText("00분")).toMatchSnapshot();
      expect(screen.getByLabelText("PM")).toMatchSnapshot();
    });

    it("초기값이 제공되면 해당 값으로 렌더링되어야 함", () => {
      const initialTime = "2024-03-20 02:30 AM";
      render(
        <TimePicker
          selectedDate={mockSelectedDate}
          setValue={mockSetValue}
          value={initialTime}
        />,
      );

      expect(screen.getByLabelText("02시")).toMatchSnapshot();
      expect(screen.getByLabelText("30분")).toMatchSnapshot();
      expect(screen.getByLabelText("AM")).toMatchSnapshot();
    });
  });

  describe("시간 선택 테스트", () => {
    it("시간 선택 시 setValue가 올바른 시간으로 호출되어야 함", () => {
      render(
        <TimePicker selectedDate={mockSelectedDate} setValue={mockSetValue} />,
      );

      fireEvent.click(screen.getByText("03"));
      fireEvent.click(screen.getByText("15"));
      fireEvent.click(screen.getByText("AM"));

      expect(mockSetValue).toHaveBeenCalledWith(
        new Date("2024-03-20 03:15:00"),
      );
    });

    it("PM 시간 선택 시 24시간 형식으로 변환되어야 함", () => {
      render(
        <TimePicker selectedDate={mockSelectedDate} setValue={mockSetValue} />,
      );

      fireEvent.click(screen.getByText("03"));
      fireEvent.click(screen.getByText("15"));
      fireEvent.click(screen.getByText("PM"));

      expect(mockSetValue).toHaveBeenCalledWith(
        new Date("2024-03-20 15:15:00"),
      );
    });

    it("12시 AM 선택 시 0시로 변환되어야 함", () => {
      render(
        <TimePicker selectedDate={mockSelectedDate} setValue={mockSetValue} />,
      );

      fireEvent.click(screen.getByText("12"));
      fireEvent.click(screen.getByText("00"));
      fireEvent.click(screen.getByText("AM"));

      expect(mockSetValue).toHaveBeenCalledWith(
        new Date("2024-03-20 00:00:00"),
      );
    });

    it("12시 PM 선택 시 12시로 유지되어야 함", () => {
      render(
        <TimePicker selectedDate={mockSelectedDate} setValue={mockSetValue} />,
      );

      fireEvent.click(screen.getByText("12"));
      fireEvent.click(screen.getByText("00"));
      fireEvent.click(screen.getByText("PM"));

      expect(mockSetValue).toHaveBeenCalledWith(
        new Date("2024-03-20 12:00:00"),
      );
    });
  });
});
