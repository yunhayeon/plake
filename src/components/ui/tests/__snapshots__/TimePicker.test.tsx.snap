// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimePicker 컴포넌트 테스트 렌더링 테스트 기본값이 제대로 렌더링되어야 함 1`] = `
<button
  aria-label="12시"
  class="flex items-center justify-center whitespace-nowrap transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-purple-600 text-white hover:bg-purple-700 rounded-xl text-base font-semibold h-9 px-4 py-2"
  role="button"
  type="button"
>
  12
</button>
`;

exports[`TimePicker 컴포넌트 테스트 렌더링 테스트 기본값이 제대로 렌더링되어야 함 2`] = `
<button
  aria-label="00분"
  class="flex items-center justify-center whitespace-nowrap transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-purple-600 text-white hover:bg-purple-700 rounded-xl text-base font-semibold h-9 px-4 py-2"
  role="button"
  type="button"
>
  00
</button>
`;

exports[`TimePicker 컴포넌트 테스트 렌더링 테스트 기본값이 제대로 렌더링되어야 함 3`] = `
<button
  aria-label="PM"
  class="flex items-center justify-center whitespace-nowrap transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-purple-600 text-white hover:bg-purple-700 rounded-xl text-base font-semibold h-9 px-4 py-2"
  role="button"
  type="button"
>
  PM
</button>
`;

exports[`TimePicker 컴포넌트 테스트 렌더링 테스트 초기값이 제공되면 해당 값으로 렌더링되어야 함 1`] = `
<button
  aria-label="02시"
  class="flex items-center justify-center whitespace-nowrap transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-purple-600 text-white hover:bg-purple-700 rounded-xl text-base font-semibold h-9 px-4 py-2"
  role="button"
  type="button"
>
  02
</button>
`;

exports[`TimePicker 컴포넌트 테스트 렌더링 테스트 초기값이 제공되면 해당 값으로 렌더링되어야 함 2`] = `
<button
  aria-label="30분"
  class="flex items-center justify-center whitespace-nowrap transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-purple-600 text-white hover:bg-purple-700 rounded-xl text-base font-semibold h-9 px-4 py-2"
  role="button"
  type="button"
>
  30
</button>
`;

exports[`TimePicker 컴포넌트 테스트 렌더링 테스트 초기값이 제공되면 해당 값으로 렌더링되어야 함 3`] = `
<button
  aria-label="AM"
  class="flex items-center justify-center whitespace-nowrap transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-purple-600 text-white hover:bg-purple-700 rounded-xl text-base font-semibold h-9 px-4 py-2"
  role="button"
  type="button"
>
  AM
</button>
`;
